omf/blockmodel/subblocks.py

```diff
@@ -0,0 +1,136 @@
+        limit on sub-block count is reached or until the sub-blocks accurately model the inputs.
+
+        This modified form of an octree also allows for stopping all cuts in one or two directions
+        before the others, so that the `subblock_count` can be (8, 4, 2) rather than (8, 8, 8).
+        All values in `subblock_count` must be a powers of two but they don't have to be equal.
+
+    "full" mode
+        The parent blocks must be either fully sub-blocked or not sub-blocked at all.
+        In other words sub-blocks must either cover the whole parent or have size (1, 1, 1).
+    """
+
+    schema = "org.omf.v2.elements.blockmodel.subblocks.regular"
+
+    subblock_count = properties.Array(
+        "The maximum number of sub-blocks inside a parent in each direction", dtype=int, shape=(3,)
+    )
+    mode = properties.StringChoice(
+        "Extra constraints on the placement of sub-blocks",
+        choices=["octree", "full"],
+        required=False,
+    )
+    parent_indices = ArrayInstanceProperty(
+        "The sub-block parent IJK indices",
+        shape=("*", 3),
+        dtype=int,
+    )
+    corners = ArrayInstanceProperty(
+        """The integer positions of the sub-block corners on the grid within their parent block""",
+        shape=("*", 6),
+        dtype=int,
+    )
+
+    @properties.validator("subblock_count")
+    def _validate_subblock_count(self, change):
+        for item in change["value"]:
+            if item < 1:
+                raise properties.ValidationError("sub-block counts must be >= 1", prop=change["name"], instance=self)
+
+    @properties.validator
+    def _validate(self):
+        _shrink_uint(self.parent_indices)
+        _shrink_uint(self.corners)
+        if self.mode == "octree":
+            for item in self.subblock_count:
+                log = np.log2(item)
+                if np.trunc(log) != log:
+                    raise properties.ValidationError(
+                        "in octree mode sub-block counts must be powers of two", prop="subblock_count", instance=self
+                    )
+
+    @property
+    def num_subblocks(self):
+        """The total number of sub-blocks."""
+        return None if self.corners is None else len(self.corners)
+
+
+class FreeformSubblocks(BaseModel):
+    """Defines free-form sub-blocks for a block model.
+
+    Sub-blocks are described by the `parent_indices` and `corners` arrays.
+
+    Each row in `parent_indices` is an IJK index on the block model grid. Each row of
+    `corners` is (i_min, j_min, k_min, i_max, j_max, k_max) all floating point, with 0.0
+    referring to the minimum side of the parent block and 1.0 the maximum side.
+    For example:
+
+    * A block with size (0.3333, 0.1, 0.5) in the corner of the parent block would be
+      (0.0, 0.0, 0.0, 0.3333, 0.1, 0.5).
+    * A sub-block covering the whole parent would be (0.0, 0.0, 0.0, 1.0, 1.0, 1.0).
```

Comment:
As discussed, if this can be made more explicit that it is a fraction of the parent that will make this clearer.